enhancements/update/dont-require-registry-during-boot-and-upgrade.md

```diff
@@ -0,0 +1,201 @@
+see-also:
+- https://issues.redhat.com/browse/OCPBUGS-13219
+- https://github.com/openshift/enhancements/pull/1481
+- https://github.com/openshift/cluster-network-operator/pull/1803
+replaces: []
+superseded-by: []
+---
+
+# Don't require registry during boot and upgrade
+
+## Summary
+
+Ensure that clusters don't require a registry server to boot or upgrade when
+all the required images have already been pulled.
+
+## Motivation
+
+Currently during reboots and upgrades clusters may need to contact the image
+registry servers, even if the images have already been pulled. This complicates
+things for clusters that are completely disconnected or that have an slow or
+unreliable connection to the image registry servers.
+
+### User Stories
+
+#### Boot without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to reboot it without requiring
+access to a registry server.
+
+### Upgrade without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to upgrade it without requiring
+access to a registry server.
+
+### Goals
+
+Ensure that clusters don't require a registry server to boot or upgrade when
+all the required images have already been pulled.
+
+### Non-Goals
+
+It is not the goal of this enhancement to add a mechanism to ensure that
+required images are available. That is the subject of the [pin and pre-load
+images](https://github.com/openshift/enhancements/pull/1481) enhancement.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster boots a node or performs an upgrade.
+
+1. All the components of the cluster ensure that if the required images have
+been already pulled they will not try to contact the registry server.
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints
+
+#### Don't use the `Always` pull policy
+
+Some OCP components currently use the `Always` image pull policy during
+upgrades. As a result, the kubelet and CRI-O will try to contact the registry
+server, even if the image is already available in the local storage of the
+cluster. This blocks upgrades and should be avoided.
+
+Most of these OCP components have been changed in the past to avoid this use of
+the `Always` pull policy. Recently the OVN pre-puller has also been changed
+(see this [bug](https://issues.redhat.com/browse/OCPBUGS-13219) for details).
+To prevent bugs like this happening in the future and make the boots and
+upgrades less fragile we should have a test that gates the OpenShift release
+and that verifies that boots and upgrades can be performed without a registry
+server. One way to ensure this is to run in CI an admission hook that
+rejects/warns about any spec that uses the `Always` pull policy.
+
+It would also be useful to have another test that scans for use of this
+`Always` pull policy in the source code.
```

Comment:
doing this via code inspection would be tough given that pod template specs can be defined in programmatic ways.  You are probably going to need to constrain it the aforementioned webhook approach

Reply:
OK, will remove this.

Reply:
Done.